File: mobile/learner_flutter/lib/services/focus_monitor_service.dart
import 'dart:async';
import 'dart:math';
import 'package:flutter/foundation.dart';
import 'package:aivo_shared/aivo_shared.dart';

/// Privacy-first focus monitoring service
/// 
/// Tracks user interactions locally and calculates focus score.
/// Only aggregated, anonymized metrics are sent to the server.
class FocusMonitorService extends ChangeNotifier {
  final String learnerId;
  final String? sessionId;
  final AivoApiClient? apiClient;
  
  // Configuration
  static const int _idleThresholdSeconds = 120; // 2 minutes
  static const int _consecutiveIncorrectThreshold = 3;
  static const double _breakSuggestionThreshold = 40.0;
  static const int _minTimeBetweenSuggestionsSeconds = 300; // 5 minutes

  // Internal state
  final List<_InteractionRecord> _interactions = [];
  final List<FocusEvent> _events = [];
  DateTime? _lastInteractionTime;
  DateTime? _lastBreakSuggestionTime;
  DateTime? _sessionStartTime;
  Timer? _idleTimer;
  
  // Tracking counters
  int _consecutiveIncorrect = 0;
  int _breaksTaken = 0;
  int _gamesPlayed = 0;
  int _totalIdleSeconds = 0;
  
  // Current state
  double _currentFocusScore = 100.0;
  bool _breakSuggested = false;
  bool _isIdle = false;
  
  // Sensory profile integration
  SensoryProfile? _sensoryProfile;

  FocusMonitorService({
    required this.learnerId,
    this.sessionId,
    this.apiClient,
  }) {
    _sessionStartTime = DateTime.now();
    _lastInteractionTime = DateTime.now();
    _startIdleTimer();
  }

  // ==================== Public Getters ====================

  double get focusScore => _currentFocusScore;
  bool get breakSuggested => _breakSuggested;
  bool get isIdle => _isIdle;
  int get breaksTaken => _breaksTaken;
  int get gamesPlayed => _gamesPlayed;
  
  /// Whether break suggestions are enabled based on sensory profile
  bool get breakSuggestionsEnabled {
    if (_sensoryProfile == null) return true;
    // Respect cognitive accommodation for avoiding popups
    if (_sensoryProfile!.cognitive.avoidPopups) return false;
    return true;
  }

  /// Break interval from sensory profile (or default)
  int get breakIntervalMinutes {
    return _sensoryProfile?.cognitive.breakIntervalMinutes ?? 20;
  }

  // ==================== Public Methods ====================

  /// Set the sensory profile for accommodation-aware behavior
  void setSensoryProfile(SensoryProfile? profile) {
    _sensoryProfile = profile;
    notifyListeners();
  }

  /// Log a user interaction
  void logInteraction(InteractionType type, {Map<String, dynamic>? data}) {
    final now = DateTime.now();
    _interactions.add(_InteractionRecord(type: type, timestamp: now, data: data));
    _lastInteractionTime = now;
    
    // Reset idle state
    if (_isIdle) {
      _isIdle = false;
      _logEvent(FocusEventType.focusRestored);
    }
    
    // Track consecutive incorrect answers
    if (type == InteractionType.incorrectAnswer) {
      _consecutiveIncorrect++;
      if (_consecutiveIncorrect >= _consecutiveIncorrectThreshold) {
        _onFrustrationDetected();
      }
    } else if (type == InteractionType.correctAnswer) {
      _consecutiveIncorrect = 0;
      _boostFocusScore(5);
    }
    
    // Restart idle timer
    _resetIdleTimer();
    
    // Recalculate focus score
    _recalculateFocusScore();
    notifyListeners();
  }

  /// Calculate current focus score (0-100)
  double calculateFocusScore() {
    return _currentFocusScore;
  }

  /// Check if a break should be suggested
  bool shouldSuggestBreak() {
    if (!breakSuggestionsEnabled) return false;
    if (_breakSuggested) return false;
    
    // Check minimum time between suggestions
    if (_lastBreakSuggestionTime != null) {
      final elapsed = DateTime.now().difference(_lastBreakSuggestionTime!);
      if (elapsed.inSeconds < _minTimeBetweenSuggestionsSeconds) {
        return false;
      }
    }
    
    // Suggest break if focus score is low
    if (_currentFocusScore < _breakSuggestionThreshold) {
      return true;
    }
    
    // Suggest break based on time interval from sensory profile
    if (_sessionStartTime != null) {
      final elapsed = DateTime.now().difference(_sessionStartTime!);
      if (elapsed.inMinutes >= breakIntervalMinutes && _breaksTaken == 0) {
        return true;
      }
    }
    
    return false;
  }

  /// Mark that a break has been suggested (to prevent repeated suggestions)
  void markBreakSuggested() {
    _breakSuggested = true;
    _lastBreakSuggestionTime = DateTime.now();
    _logEvent(FocusEventType.distractionDetected, focusScore: _currentFocusScore);
    notifyListeners();
  }

  /// Called when user starts a break
  void startBreak() {
    _logEvent(FocusEventType.breakStarted, focusScore: _currentFocusScore);
    notifyListeners();
  }

  /// Called when user completes a break
  void completeBreak({int? durationSeconds, String? gameId}) {
    _breaksTaken++;
    _breakSuggested = false;
    
    // Restore focus after break
    _currentFocusScore = min(100.0, _currentFocusScore + 30);
    _consecutiveIncorrect = 0;
    
    _logEvent(
      FocusEventType.breakCompleted,
      focusScore: _currentFocusScore,
      durationSeconds: durationSeconds,
      gameId: gameId,
    );
    
    notifyListeners();
  }

  /// Called when user plays a game
  void logGamePlayed(String gameId, int durationSeconds) {
    _gamesPlayed++;
    _logEvent(
      FocusEventType.gamePlayed,
      gameId: gameId,
      durationSeconds: durationSeconds,
    );
    notifyListeners();
  }

  /// Dismiss break suggestion without taking a break
  void dismissBreakSuggestion() {
    _breakSuggested = false;
    _lastBreakSuggestionTime = DateTime.now();
    notifyListeners();
  }

  /// Get aggregated metrics for server sync (privacy-safe)
  FocusMetrics getAggregatedMetrics() {
    final now = DateTime.now();
    final activeSeconds = _interactions.length > 0
        ? _interactions.last.timestamp.difference(_sessionStartTime!).inSeconds - _totalIdleSeconds
        : 0;
    
    // Calculate average focus score from events
    final focusEvents = _events.where((e) => e.focusScore != null).toList();
    final avgScore = focusEvents.isNotEmpty
        ? focusEvents.map((e) => e.focusScore!).reduce((a, b) => a + b) / focusEvents.length
        : _currentFocusScore;
    
    return FocusMetrics(
      learnerId: learnerId,
      sessionId: sessionId ?? 'unknown',
      averageFocusScore: avgScore,
      breaksTaken: _breaksTaken,
      gamesPlayed: _gamesPlayed,
      totalActiveSeconds: activeSeconds,
      totalIdleSeconds: _totalIdleSeconds,
      periodStart: _sessionStartTime!,
      periodEnd: now,
    );
  }

  /// Sync aggregated metrics to server
  Future<void> syncMetrics() async {
    if (apiClient == null) return;
    
    try {
      final metrics = getAggregatedMetrics();
      // API call would go here
      // await apiClient!.syncFocusMetrics(metrics);
    } catch (e) {
      // Fail silently - metrics are not critical
      debugPrint('Failed to sync focus metrics: $e');
    }
  }

  /// Reset the monitor for a new session
  void reset() {
    _interactions.clear();
    _events.clear();
    _sessionStartTime = DateTime.now();
    _lastInteractionTime = DateTime.now();
    _lastBreakSuggestionTime = null;
    _consecutiveIncorrect = 0;
    _breaksTaken = 0;
    _gamesPlayed = 0;
    _totalIdleSeconds = 0;
    _currentFocusScore = 100.0;
    _breakSuggested = false;
    _isIdle = false;
    notifyListeners();
  }

  @override
  void dispose() {
    _idleTimer?.cancel();
    super.dispose();
  }

  // ==================== Private Methods ====================

  void _startIdleTimer() {
    _idleTimer?.cancel();
    _idleTimer = Timer.periodic(const Duration(seconds: 10), (_) {
      _checkIdle();
    });
  }

  void _resetIdleTimer() {
    _idleTimer?.cancel();
    _startIdleTimer();
  }

  void _checkIdle() {
    if (_lastInteractionTime == null) return;
    
    final elapsed = DateTime.now().difference(_lastInteractionTime!);
    if (elapsed.inSeconds >= _idleThresholdSeconds && !_isIdle) {
      _onIdleDetected();
    }
    
    if (_isIdle) {
      _totalIdleSeconds += 10;
    }
  }

  void _onIdleDetected() {
    _isIdle = true;
    _currentFocusScore = max(0, _currentFocusScore - 15);
    _logEvent(FocusEventType.idleDetected, focusScore: _currentFocusScore);
    notifyListeners();
  }

  void _onFrustrationDetected() {
    _currentFocusScore = max(0, _currentFocusScore - 20);
    _logEvent(FocusEventType.distractionDetected, focusScore: _currentFocusScore);
    notifyListeners();
  }

  void _boostFocusScore(double amount) {
    _currentFocusScore = min(100, _currentFocusScore + amount);
  }

  void _recalculateFocusScore() {
    // Get recent interactions (last 2 minutes)
    final cutoff = DateTime.now().subtract(const Duration(minutes: 2));
    final recentInteractions = _interactions.where(
      (i) => i.timestamp.isAfter(cutoff),
    ).toList();
    
    if (recentInteractions.isEmpty) {
      _currentFocusScore = max(0, _currentFocusScore - 5);
      return;
    }
    
    // Calculate score based on interaction patterns
    double score = _currentFocusScore;
    
    // Factor 1: Interaction frequency (more is better, up to a point)
    final interactionsPerMinute = recentInteractions.length / 2.0;
    if (interactionsPerMinute < 1) {
      score -= 5;
    } else if (interactionsPerMinute > 10) {
      // Too many interactions might indicate frustration
      score -= 10;
    } else {
      score += 2;
    }
    
    // Factor 2: Correct vs incorrect answers
    final correctCount = recentInteractions
        .where((i) => i.type == InteractionType.correctAnswer)
        .length;
    final incorrectCount = recentInteractions
        .where((i) => i.type == InteractionType.incorrectAnswer)
        .length;
    
    if (incorrectCount > 0 && correctCount == 0) {
      score -= 10;
    } else if (correctCount > incorrectCount) {
      score += 5;
    }
    
    // Factor 3: Check for erratic patterns (rapid taps)
    final tapInteractions = recentInteractions
        .where((i) => i.type == InteractionType.tap)
        .toList();
    if (tapInteractions.length > 20) {
      // Too many taps in 2 minutes suggests frustration
      score -= 15;
    }
    
    _currentFocusScore = score.clamp(0, 100);
  }

  void _logEvent(
    FocusEventType eventType, {
    double? focusScore,
    String? gameId,
    int? durationSeconds,
  }) {
    _events.add(FocusEvent(
      learnerId: learnerId,
      eventType: eventType,
      focusScore: focusScore,
      gameId: gameId,
      durationSeconds: durationSeconds,
      timestamp: DateTime.now(),
    ));
  }
}

/// Internal record of an interaction
class _InteractionRecord {
  final InteractionType type;
  final DateTime timestamp;
  final Map<String, dynamic>? data;

  _InteractionRecord({
    required this.type,
    required this.timestamp,
    this.data,
  });
}

// ==================== Pre-built Focus Break Games ====================

/// Factory for creating pre-built focus break games
class FocusBreakGames {
  static final Random _random = Random();

  /// Get all available games
  static List<FocusBreakGame> getAllGames() {
    return [
      memoryGame(),
      quickMathGame(),
      wordScrambleGame(),
      movementBreak(),
      breathingExercise(),
    ];
  }

  /// Memory card matching game
  static FocusBreakGame memoryGame() {
    final emojis = ['🐶', '🐱', '🐭', '🐹', '🐰', '🦊', '🐻', '🐼'];
    final pairs = <Map<String, dynamic>>[];
    
    for (var i = 0; i < emojis.length; i++) {
      pairs.add({'id': i, 'emoji': emojis[i]});
    }
    
    return FocusBreakGame(
      id: 'memory-1',
      type: FocusBreakGameType.memory,
      name: 'Memory Match',
      description: 'Find matching pairs of animals!',
      durationSeconds: 120,
      gameData: {
        'pairs': pairs,
        'gridSize': 4,
      },
    );
  }

  /// Quick arithmetic game
  static FocusBreakGame quickMathGame() {
    final problems = <Map<String, dynamic>>[];
    
    for (var i = 0; i < 10; i++) {
      final a = _random.nextInt(10) + 1;
      final b = _random.nextInt(10) + 1;
      final isAddition = _random.nextBool();
      
      final question = isAddition ? '$a + $b' : '${a + b} - $a';
      final correct = isAddition ? a + b : b;
      
      // Generate wrong options
      final options = <int>[correct];
      while (options.length < 4) {
        final wrong = correct + _random.nextInt(5) - 2;
        if (wrong != correct && wrong > 0 && !options.contains(wrong)) {
          options.add(wrong);
        }
      }
      options.shuffle();
      
      problems.add({
        'question': question,
        'options': options,
        'correct': options.indexOf(correct),
      });
    }
    
    return FocusBreakGame(
      id: 'math-1',
      type: FocusBreakGameType.quickMath,
      name: 'Quick Math',
      description: 'Solve as many as you can in 30 seconds!',
      durationSeconds: 30,
      gameData: {
        'problems': problems,
      },
    );
  }

  /// Word unscrambling game
  static FocusBreakGame wordScrambleGame() {
    final words = [
      'APPLE', 'HAPPY', 'MUSIC', 'DANCE', 'SMILE',
      'BRAIN', 'LEARN', 'FOCUS', 'SMART', 'THINK',
    ];
    
    final puzzles = words.map((word) {
      final letters = word.split('');
      letters.shuffle();
      return {
        'scrambled': letters.join(''),
        'answer': word,
        'hint': '${word[0]}...',
      };
    }).toList();
    
    return FocusBreakGame(
      id: 'scramble-1',
      type: FocusBreakGameType.wordScramble,
      name: 'Word Scramble',
      description: 'Unscramble the letters to find the word!',
      durationSeconds: 60,
      gameData: {
        'puzzles': puzzles,
      },
    );
  }

  /// Movement/exercise break
  static FocusBreakGame movementBreak() {
    final exercises = [
      {'name': 'Jumping Jacks', 'count': 10, 'emoji': '🤸'},
      {'name': 'Arm Circles', 'count': 10, 'emoji': '🙆'},
      {'name': 'Toe Touches', 'count': 5, 'emoji': '🧘'},
      {'name': 'March in Place', 'count': 20, 'emoji': '🚶'},
      {'name': 'Shoulder Shrugs', 'count': 10, 'emoji': '💪'},
    ];
    
    return FocusBreakGame(
      id: 'movement-1',
      type: FocusBreakGameType.movement,
      name: 'Movement Break',
      description: 'Get your body moving!',
      durationSeconds: 90,
      gameData: {
        'exercises': exercises,
      },
    );
  }

  /// Breathing/calming exercise
  static FocusBreakGame breathingExercise() {
    return FocusBreakGame(
      id: 'breathing-1',
      type: FocusBreakGameType.breathing,
      name: 'Calm Breathing',
      description: 'Follow the circle to calm your mind',
      durationSeconds: 60,
      gameData: {
        'pattern': {
          'inhaleSeconds': 4,
          'holdSeconds': 4,
          'exhaleSeconds': 4,
          'cycles': 4,
        },
      },
    );
  }

  /// Get a random game
  static FocusBreakGame randomGame() {
    final games = getAllGames();
    return games[_random.nextInt(games.length)];
  }

  /// Get a game by type
  static FocusBreakGame? getByType(FocusBreakGameType type) {
    try {
      return getAllGames().firstWhere((g) => g.type == type);
    } catch (_) {
      return null;
    }
  }
}

File: apps/parent-teacher-web/services/FocusMonitorService.ts
/**
 * Focus Monitor Service for Web
 * Privacy-first focus monitoring with local interaction tracking
 *
 * Ported from Flutter: mobile/learner_flutter/lib/services/focus_monitor_service.dart
 */

// ==================== Types ====================

export enum InteractionType {
  Tap = 'tap',
  CorrectAnswer = 'correctAnswer',
  IncorrectAnswer = 'incorrectAnswer',
  PageView = 'pageView',
  ScrollAction = 'scrollAction',
  FormInput = 'formInput',
  ButtonClick = 'buttonClick',
}

export enum FocusEventType {
  FocusRestored = 'focusRestored',
  IdleDetected = 'idleDetected',
  DistractionDetected = 'distractionDetected',
  BreakStarted = 'breakStarted',
  BreakCompleted = 'breakCompleted',
  GamePlayed = 'gamePlayed',
}

export interface FocusEvent {
  learnerId: string;
  eventType: FocusEventType;
  focusScore?: number;
  gameId?: string;
  durationSeconds?: number;
  timestamp: Date;
}

export interface FocusMetrics {
  learnerId: string;
  sessionId: string;
  averageFocusScore: number;
  breaksTaken: number;
  gamesPlayed: number;
  totalActiveSeconds: number;
  totalIdleSeconds: number;
  periodStart: Date;
  periodEnd: Date;
}

export interface SensoryProfileCognitive {
  avoidPopups?: boolean;
  breakIntervalMinutes?: number;
}

export interface SensoryProfile {
  cognitive?: SensoryProfileCognitive;
}

interface InteractionRecord {
  type: InteractionType;
  timestamp: Date;
  data?: Record<string, unknown>;
}

type FocusChangeListener = (score: number, breakSuggested: boolean) => void;

// ==================== Focus Monitor Service ====================

export class FocusMonitorService {
  private readonly learnerId: string;
  private readonly sessionId: string;

  // Configuration
  private static readonly IDLE_THRESHOLD_SECONDS = 120; // 2 minutes
  private static readonly CONSECUTIVE_INCORRECT_THRESHOLD = 3;
  private static readonly BREAK_SUGGESTION_THRESHOLD = 40.0;
  private static readonly MIN_TIME_BETWEEN_SUGGESTIONS_SECONDS = 300; // 5 minutes

  // Internal state
  private interactions: InteractionRecord[] = [];
  private events: FocusEvent[] = [];
  private lastInteractionTime: Date | null = null;
  private lastBreakSuggestionTime: Date | null = null;
  private sessionStartTime: Date;
  private idleTimerId: ReturnType<typeof setInterval> | null = null;

  // Tracking counters
  private consecutiveIncorrect = 0;
  private _breaksTaken = 0;
  private _gamesPlayed = 0;
  private totalIdleSeconds = 0;

  // Current state
  private _currentFocusScore = 100.0;
  private _breakSuggested = false;
  private _isIdle = false;

  // Sensory profile integration
  private sensoryProfile: SensoryProfile | null = null;

  // Listeners
  private listeners: FocusChangeListener[] = [];

  constructor(learnerId: string, sessionId?: string) {
    this.learnerId = learnerId;
    this.sessionId = sessionId ?? crypto.randomUUID();
    this.sessionStartTime = new Date();
    this.lastInteractionTime = new Date();
    this.startIdleTimer();
  }

  // ==================== Public Getters ====================

  get focusScore(): number {
    return this._currentFocusScore;
  }

  get breakSuggested(): boolean {
    return this._breakSuggested;
  }

  get isIdle(): boolean {
    return this._isIdle;
  }

  get breaksTaken(): number {
    return this._breaksTaken;
  }

  get gamesPlayed(): number {
    return this._gamesPlayed;
  }

  get breakSuggestionsEnabled(): boolean {
    if (!this.sensoryProfile) return true;
    if (this.sensoryProfile.cognitive?.avoidPopups) return false;
    return true;
  }

  get breakIntervalMinutes(): number {
    return this.sensoryProfile?.cognitive?.breakIntervalMinutes ?? 20;
  }

  // ==================== Public Methods ====================

  /**
   * Add a listener for focus changes
   */
  addListener(listener: FocusChangeListener): void {
    this.listeners.push(listener);
  }

  /**
   * Remove a listener
   */
  removeListener(listener: FocusChangeListener): void {
    const index = this.listeners.indexOf(listener);
    if (index > -1) {
      this.listeners.splice(index, 1);
    }
  }

  /**
   * Notify all listeners of changes
   */
  private notifyListeners(): void {
    for (const listener of this.listeners) {
      listener(this._currentFocusScore, this._breakSuggested);
    }
  }

  /**
   * Set the sensory profile for accommodation-aware behavior
   */
  setSensoryProfile(profile: SensoryProfile | null): void {
    this.sensoryProfile = profile;
    this.notifyListeners();
  }

  /**
   * Log a user interaction
   */
  logInteraction(type: InteractionType, data?: Record<string, unknown>): void {
    const now = new Date();
    this.interactions.push({ type, timestamp: now, data });
    this.lastInteractionTime = now;

    // Reset idle state
    if (this._isIdle) {
      this._isIdle = false;
      this.logEvent(FocusEventType.FocusRestored);
    }

    // Track consecutive incorrect answers
    if (type === InteractionType.IncorrectAnswer) {
      this.consecutiveIncorrect++;
      if (this.consecutiveIncorrect >= FocusMonitorService.CONSECUTIVE_INCORRECT_THRESHOLD) {
        this.onFrustrationDetected();
      }
    } else if (type === InteractionType.CorrectAnswer) {
      this.consecutiveIncorrect = 0;
      this.boostFocusScore(5);
    }

    // Restart idle timer
    this.resetIdleTimer();

    // Recalculate focus score
    this.recalculateFocusScore();
    this.notifyListeners();
  }

  /**
   * Calculate current focus score (0-100)
   */
  calculateFocusScore(): number {
    return this._currentFocusScore;
  }

  /**
   * Check if a break should be suggested
   */
  shouldSuggestBreak(): boolean {
    if (!this.breakSuggestionsEnabled) return false;
    if (this._breakSuggested) return false;

    // Check minimum time between suggestions
    if (this.lastBreakSuggestionTime) {
      const elapsed = (new Date().getTime() - this.lastBreakSuggestionTime.getTime()) / 1000;
      if (elapsed < FocusMonitorService.MIN_TIME_BETWEEN_SUGGESTIONS_SECONDS) {
        return false;
      }
    }

    // Suggest break if focus score is low
    if (this._currentFocusScore < FocusMonitorService.BREAK_SUGGESTION_THRESHOLD) {
      return true;
    }

    // Suggest break based on time interval from sensory profile
    const elapsed = (new Date().getTime() - this.sessionStartTime.getTime()) / 60000;
    if (elapsed >= this.breakIntervalMinutes && this._breaksTaken === 0) {
      return true;
    }

    return false;
  }

  /**
   * Mark that a break has been suggested
   */
  markBreakSuggested(): void {
    this._breakSuggested = true;
    this.lastBreakSuggestionTime = new Date();
    this.logEvent(FocusEventType.DistractionDetected, this._currentFocusScore);
    this.notifyListeners();
  }

  /**
   * Called when user starts a break
   */
  startBreak(): void {
    this.logEvent(FocusEventType.BreakStarted, this._currentFocusScore);
    this.notifyListeners();
  }

  /**
   * Called when user completes a break
   */
  completeBreak(durationSeconds?: number, gameId?: string): void {
    this._breaksTaken++;
    this._breakSuggested = false;

    // Restore focus after break
    this._currentFocusScore = Math.min(100.0, this._currentFocusScore + 30);
    this.consecutiveIncorrect = 0;

    this.logEvent(FocusEventType.BreakCompleted, this._currentFocusScore, durationSeconds, gameId);

    this.notifyListeners();
  }

  /**
   * Called when user plays a game
   */
  logGamePlayed(gameId: string, durationSeconds: number): void {
    this._gamesPlayed++;
    this.logEvent(FocusEventType.GamePlayed, undefined, durationSeconds, gameId);
    this.notifyListeners();
  }

  /**
   * Dismiss break suggestion without taking a break
   */
  dismissBreakSuggestion(): void {
    this._breakSuggested = false;
    this.lastBreakSuggestionTime = new Date();
    this.notifyListeners();
  }

  /**
   * Get aggregated metrics for server sync (privacy-safe)
   */
  getAggregatedMetrics(): FocusMetrics {
    const now = new Date();
    const activeSeconds =
      this.interactions.length > 0
        ? (this.interactions[this.interactions.length - 1].timestamp.getTime() -
            this.sessionStartTime.getTime()) /
            1000 -
          this.totalIdleSeconds
        : 0;

    // Calculate average focus score from events
    const focusEvents = this.events.filter((e) => e.focusScore !== undefined);
    const avgScore =
      focusEvents.length > 0
        ? focusEvents.reduce((sum, e) => sum + (e.focusScore ?? 0), 0) / focusEvents.length
        : this._currentFocusScore;

    return {
      learnerId: this.learnerId,
      sessionId: this.sessionId,
      averageFocusScore: avgScore,
      breaksTaken: this._breaksTaken,
      gamesPlayed: this._gamesPlayed,
      totalActiveSeconds: Math.max(0, activeSeconds),
      totalIdleSeconds: this.totalIdleSeconds,
      periodStart: this.sessionStartTime,
      periodEnd: now,
    };
  }

  /**
   * Sync aggregated metrics to server
   */
  async syncMetrics(): Promise<void> {
    try {
      const metrics = this.getAggregatedMetrics();
      // API call would go here
      await fetch('/api/focus/metrics', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(metrics),
      });
    } catch (e) {
      // Fail silently - metrics are not critical
      console.warn('Failed to sync focus metrics:', e);
    }
  }

  /**
   * Reset the monitor for a new session
   */
  reset(): void {
    this.interactions = [];
    this.events = [];
    this.sessionStartTime = new Date();
    this.lastInteractionTime = new Date();
    this.lastBreakSuggestionTime = null;
    this.consecutiveIncorrect = 0;
    this._breaksTaken = 0;
    this._gamesPlayed = 0;
    this.totalIdleSeconds = 0;
    this._currentFocusScore = 100.0;
    this._breakSuggested = false;
    this._isIdle = false;
    this.notifyListeners();
  }

  /**
   * Clean up resources
   */
  dispose(): void {
    if (this.idleTimerId) {
      clearInterval(this.idleTimerId);
      this.idleTimerId = null;
    }
    this.listeners = [];
  }

  // ==================== Private Methods ====================

  private startIdleTimer(): void {
    if (this.idleTimerId) {
      clearInterval(this.idleTimerId);
    }
    this.idleTimerId = setInterval(() => this.checkIdle(), 10000);
  }

  private resetIdleTimer(): void {
    this.startIdleTimer();
  }

  private checkIdle(): void {
    if (!this.lastInteractionTime) return;

    const elapsed = (new Date().getTime() - this.lastInteractionTime.getTime()) / 1000;
    if (elapsed >= FocusMonitorService.IDLE_THRESHOLD_SECONDS && !this._isIdle) {
      this.onIdleDetected();
    }

    if (this._isIdle) {
      this.totalIdleSeconds += 10;
    }
  }

  private onIdleDetected(): void {
    this._isIdle = true;
    this._currentFocusScore = Math.max(0, this._currentFocusScore - 15);
    this.logEvent(FocusEventType.IdleDetected, this._currentFocusScore);
    this.notifyListeners();
  }

  private onFrustrationDetected(): void {
    this._currentFocusScore = Math.max(0, this._currentFocusScore - 20);
    this.logEvent(FocusEventType.DistractionDetected, this._currentFocusScore);
    this.notifyListeners();
  }

  private boostFocusScore(amount: number): void {
    this._currentFocusScore = Math.min(100, this._currentFocusScore + amount);
  }

  private recalculateFocusScore(): void {
    // Get recent interactions (last 2 minutes)
    const cutoff = new Date(Date.now() - 2 * 60 * 1000);
    const recentInteractions = this.interactions.filter((i) => i.timestamp > cutoff);

    if (recentInteractions.length === 0) {
      this._currentFocusScore = Math.max(0, this._currentFocusScore - 5);
      return;
    }

    // Calculate score based on interaction patterns
    let score = this._currentFocusScore;

    // Factor 1: Interaction frequency (more is better, up to a point)
    const interactionsPerMinute = recentInteractions.length / 2.0;
    if (interactionsPerMinute < 1) {
      score -= 5;
    } else if (interactionsPerMinute > 10) {
      // Too many interactions might indicate frustration
      score -= 10;
    } else {
      score += 2;
    }

    // Factor 2: Correct vs incorrect answers
    const correctCount = recentInteractions.filter(
      (i) => i.type === InteractionType.CorrectAnswer,
    ).length;
    const incorrectCount = recentInteractions.filter(
      (i) => i.type === InteractionType.IncorrectAnswer,
    ).length;

    if (incorrectCount > 0 && correctCount === 0) {
      score -= 10;
    } else if (correctCount > incorrectCount) {
      score += 5;
    }

    // Factor 3: Check for erratic patterns (rapid taps)
    const tapInteractions = recentInteractions.filter((i) => i.type === InteractionType.Tap);
    if (tapInteractions.length > 20) {
      // Too many taps in 2 minutes suggests frustration
      score -= 15;
    }

    this._currentFocusScore = Math.max(0, Math.min(100, score));
  }

  private logEvent(
    eventType: FocusEventType,
    focusScore?: number,
    durationSeconds?: number,
    gameId?: string,
  ): void {
    this.events.push({
      learnerId: this.learnerId,
      eventType,
      focusScore,
      gameId,
      durationSeconds,
      timestamp: new Date(),
    });
  }
}

// ==================== React Hook ====================

import { useState, useEffect, useCallback, useRef } from 'react';

/**
 * React hook for using FocusMonitorService
 */
export function useFocusMonitor(learnerId: string, sessionId?: string) {
  const serviceRef = useRef<FocusMonitorService | null>(null);
  const [focusScore, setFocusScore] = useState(100);
  const [breakSuggested, setBreakSuggested] = useState(false);
  const [isIdle, setIsIdle] = useState(false);

  useEffect(() => {
    const service = new FocusMonitorService(learnerId, sessionId);
    serviceRef.current = service;

    const listener = (score: number, breakSuggested: boolean) => {
      setFocusScore(score);
      setBreakSuggested(breakSuggested);
      setIsIdle(service.isIdle);
    };

    service.addListener(listener);

    return () => {
      service.dispose();
    };
  }, [learnerId, sessionId]);

  const logInteraction = useCallback((type: InteractionType, data?: Record<string, unknown>) => {
    serviceRef.current?.logInteraction(type, data);
  }, []);

  const completeBreak = useCallback((durationSeconds?: number, gameId?: string) => {
    serviceRef.current?.completeBreak(durationSeconds, gameId);
  }, []);

  const dismissBreakSuggestion = useCallback(() => {
    serviceRef.current?.dismissBreakSuggestion();
  }, []);

  const getMetrics = useCallback(() => {
    return serviceRef.current?.getAggregatedMetrics();
  }, []);

  const reset = useCallback(() => {
    serviceRef.current?.reset();
  }, []);

  return {
    focusScore,
    breakSuggested,
    isIdle,
    logInteraction,
    completeBreak,
    dismissBreakSuggestion,
    getMetrics,
    reset,
    service: serviceRef.current,
  };
}

export default FocusMonitorService;
